File: agent/app/modules/storygraph/vision_review_input.py
from __future__ import annotations

from typing import Literal

from pydantic import Field, model_validator

from app.modules.storygraph.reference_brief_contract import (
    ReferenceBriefCandidate,
    ReferenceBriefInput,
)
from app.modules.storygraph.reference_plan_contract import ReferencePlanPurposeProfile
from app.modules.storygraph.scene_analysis_candidates import StrictSceneAnalysisModel
from app.modules.storygraph.vision_review_attachments import (
    VisionReviewMediaAttachment,
    validate_vision_review_attachments,
)
from app.modules.storygraph.vision_review_contract import Digest, VisionReviewSubject
from app.modules.storygraph.visual_foundation_contract import (
    FIDELITY_INVARIANTS,
    CreativeFillProposal,
    VisualFoundationPolicy,
    VisualWorldConflict,
    WorldAdaptationProposal,
)
from app.protocol.canonical import production_canonical_hash, production_canonical_json_text


class VisionReviewVisualGrammar(StrictSceneAnalysisModel):
    medium: str
    realism: str
    shape_language: str
    proportion: str
    palette: str
    linework: str
    texture: str
    material_rendering: str
    lighting: str
    contrast: str
    composition: str
    camera: str
    negative_constraints: list[str]

    @model_validator(mode="after")
    def validate_grammar(self) -> VisionReviewVisualGrammar:
        for value in (
            self.medium,
            self.realism,
            self.shape_language,
            self.proportion,
            self.palette,
            self.linework,
            self.texture,
            self.material_rendering,
            self.lighting,
            self.contrast,
            self.composition,
            self.camera,
        ):
            _validate_text(value)
        _validate_rules(self.negative_constraints)
        return self


class VisionReviewPolicyRef(StrictSceneAnalysisModel):
    owner: str
    key: str
    content_hash: Digest

    @model_validator(mode="after")
    def validate_reference(self) -> VisionReviewPolicyRef:
        _validate_text(self.owner)
        _validate_text(self.key)
        return self


class VisionReviewVisualContext(StrictSceneAnalysisModel):
    application_mode: Literal["faithful", "world_adaptation"]
    production_world_owner_set_hash: Digest
    visual_grammar: VisionReviewVisualGrammar
    style_policy: VisualFoundationPolicy
    fidelity_invariants: list[str]
    world_adaptations: list[WorldAdaptationProposal]
    world_conflicts: list[VisualWorldConflict]
    creative_fill_proposals: list[CreativeFillProposal]
    purpose_profile: ReferencePlanPurposeProfile
    qc_policy_ref: VisionReviewPolicyRef

    @model_validator(mode="after")
    def validate_context(self) -> VisionReviewVisualContext:
        if self.fidelity_invariants != FIDELITY_INVARIANTS:
            raise ValueError("Vision Review cannot relax fidelity invariants")
        if self.application_mode == "faithful" and self.world_adaptations:
            raise ValueError("faithful Vision Review cannot adapt world facts")
        _validate_rules(self.purpose_profile.design_focus)
        _validate_rules(self.purpose_profile.forbidden_changes)
        for keys in (
            [value.mapping_key for value in self.world_adaptations],
            [value.conflict_key for value in self.world_conflicts],
            [value.gap_key for value in self.creative_fill_proposals],
        ):
            if keys != sorted(set(keys)):
                raise ValueError("Vision Review policy proposals must be ordered and unique")
        for rules in (
            self.style_policy.palette_rules,
            self.style_policy.material_rules,
            self.style_policy.lighting_rules,
            self.style_policy.camera_rules,
        ):
            for rule in rules:
                _validate_text(rule)
                if len(rule.encode("utf-8")) > 800:
                    raise ValueError("Vision Review style policy rule exceeds its byte budget")
        for value in self.world_adaptations:
            if len(value.design_value.encode("utf-8")) > 1200:
                raise ValueError("Vision Review adaptation exceeds its byte budget")
        for value in self.world_conflicts:
            if len(value.summary.encode("utf-8")) > 800:
                raise ValueError("Vision Review conflict exceeds its byte budget")
        for value in self.creative_fill_proposals:
            if len(value.proposal.encode("utf-8")) > 1200:
                raise ValueError("Vision Review proposal exceeds its byte budget")
        return self


class VisionReviewAdmissionPolicyRef(StrictSceneAnalysisModel):
    contract_id: Literal["reference-bundle-purpose-admission"]
    content_hash: Digest


class VisionReviewAdmission(StrictSceneAnalysisModel):
    policy_ref: VisionReviewAdmissionPolicyRef
    internal_review_ready: bool = Field(strict=True)
    selection_ready: bool = Field(strict=True)
    publication_ready: bool = Field(strict=True)
    internal_review_blockers: list[str]
    formal_use_blockers: list[str]

    @model_validator(mode="after")
    def validate_material_admission(self) -> VisionReviewAdmission:
        policy_hash = production_canonical_hash(
            {
                "formal_use_requires": ["rights_assessment", "vision_review"],
                "internal_review_requires": ["complete_bundle", "technical_qc"],
                "rights_not_assessed": "internal_review_only",
            }
        )
        if (
            self.policy_ref.content_hash != policy_hash
            or not self.internal_review_ready
            or self.selection_ready
            or self.publication_ready
            or self.internal_review_blockers
            or self.formal_use_blockers != ["rights_not_assessed", "vision_review_required"]
        ):
            raise ValueError("Vision Review material admission is invalid")
        return self


class VisionReviewInput(StrictSceneAnalysisModel):
    """Closed base-wave input; never a substitute for current Owner authorization."""

    contract_id: Literal["vision-review-input-production"]
    subject: VisionReviewSubject
    brief_input: ReferenceBriefInput
    brief_candidate: ReferenceBriefCandidate
    brief_content_hash: Digest
    visual_context: VisionReviewVisualContext
    admission: VisionReviewAdmission
    attachments: list[VisionReviewMediaAttachment] = Field(min_length=1, max_length=4)

    @model_validator(mode="after")
    def validate_input(self) -> VisionReviewInput:
        self.brief_candidate.validate_for(self.brief_input)
        validate_vision_review_attachments(self.subject, self.attachments)
        if (
            self.subject.target_kind
            not in {"character_identity_anchor", "location_board", "prop_sheet"}
            or self.brief_input.dependency_selections
        ):
            raise ValueError("Vision Review dependency assets are not ready")
        if (
            self.subject.workspace_id != str(self.brief_input.workspace_id)
            or self.subject.project_id != str(self.brief_input.project_id)
            or self.subject.target_kind != self.brief_input.target_kind
            or self.subject.generation_round != 1
            or self.visual_context.purpose_profile.target_kind != self.subject.target_kind
        ):
            raise ValueError("Vision Review Brief or visual context scope has drifted")
        if (
            production_canonical_hash(self.brief_candidate.model_dump(mode="json"))
            != self.brief_content_hash
        ):
            raise ValueError("Vision Review Brief content has drifted")
        material = self.model_dump(mode="json")
        del material["subject"]["input_hash"]
        if production_canonical_hash(material) != self.subject.input_hash:
            raise ValueError("Vision Review input hash has drifted")
        return self


def decode_vision_review_input(raw: str | bytes) -> VisionReviewInput:
    wire = production_canonical_json_text(raw.encode("utf-8") if isinstance(raw, str) else raw)
    value = VisionReviewInput.model_validate_json(wire)
    typed = production_canonical_json_text(value.model_dump_json().encode("utf-8"))
    if typed != wire:
        raise ValueError("Vision Review input wire shape is incomplete")
    return value


def _validate_text(value: str) -> None:
    if not value or value.strip() != value:
        raise ValueError("Vision Review requires canonical visual text")


def _validate_rules(values: list[str]) -> None:
    if not values or values != sorted(set(values)):
        raise ValueError("Vision Review visual rules must be explicit and ordered")
    for value in values:
        _validate_text(value)

File: backend/internal/agent/contract/vision_review_input.go
package contract

import (
	"bytes"
	"encoding/json"
	"errors"
	"reflect"
	"slices"
	"strings"

	generation "github.com/StephenQiu30/lanverse/backend/internal/generation/domain"
	"github.com/StephenQiu30/lanverse/backend/internal/platform/canonical"
)

const VisionReviewInputContractID = "vision-review-input-production"

// VisionReviewVisualGrammar is the complete review-facing Preset grammar, not
// the smaller design-stage snapshot or an independently writable Preset fact.
type VisionReviewVisualGrammar struct {
	Medium              string   `json:"medium"`
	Realism             string   `json:"realism"`
	ShapeLanguage       string   `json:"shape_language"`
	Proportion          string   `json:"proportion"`
	Palette             string   `json:"palette"`
	Linework            string   `json:"linework"`
	Texture             string   `json:"texture"`
	MaterialRendering   string   `json:"material_rendering"`
	Lighting            string   `json:"lighting"`
	Contrast            string   `json:"contrast"`
	Composition         string   `json:"composition"`
	Camera              string   `json:"camera"`
	NegativeConstraints []string `json:"negative_constraints"`
}

type VisionReviewPolicyRef struct {
	Owner       string `json:"owner"`
	Key         string `json:"key"`
	ContentHash string `json:"content_hash"`
}

type VisionReviewVisualContext struct {
	ApplicationMode             string                      `json:"application_mode"`
	ProductionWorldOwnerSetHash string                      `json:"production_world_owner_set_hash"`
	VisualGrammar               VisionReviewVisualGrammar   `json:"visual_grammar"`
	StylePolicy                 VisualFoundationPolicy      `json:"style_policy"`
	FidelityInvariants          []string                    `json:"fidelity_invariants"`
	WorldAdaptations            []WorldAdaptationProposal   `json:"world_adaptations"`
	WorldConflicts              []VisualWorldConflict       `json:"world_conflicts"`
	CreativeFillProposals       []CreativeFillProposal      `json:"creative_fill_proposals"`
	PurposeProfile              ReferencePlanPurposeProfile `json:"purpose_profile"`
	QCPolicyRef                 VisionReviewPolicyRef       `json:"qc_policy_ref"`
}

type VisionReviewInput struct {
	ContractID       string                              `json:"contract_id"`
	Subject          VisionReviewSubject                 `json:"subject"`
	BriefInput       ReferenceBriefInput                 `json:"brief_input"`
	BriefCandidate   ReferenceBriefCandidate             `json:"brief_candidate"`
	BriefContentHash string                              `json:"brief_content_hash"`
	VisualContext    VisionReviewVisualContext           `json:"visual_context"`
	Admission        generation.ReferenceBundleAdmission `json:"admission"`
	Attachments      []VisionReviewMediaAttachment       `json:"attachments"`
}

// BuildVisionReviewInput seals a typed projection. Only the Backend compiler
// can establish its Owner provenance; sealing never establishes authorization.
func BuildVisionReviewInput(input VisionReviewInput) (VisionReviewInput, error) {
	input.ContractID = VisionReviewInputContractID
	input.Subject.InputHash = strings.Repeat("0", 64)
	if err := input.validateShape(); err != nil {
		return VisionReviewInput{}, err
	}
	hash, err := visionReviewInputHash(input)
	if err != nil {
		return VisionReviewInput{}, err
	}
	input.Subject.InputHash = hash
	// Detach all nested mutable slices and maps from the caller's Owner snapshots.
	raw, err := json.Marshal(input)
	if err != nil {
		return VisionReviewInput{}, err
	}
	return DecodeVisionReviewInput(raw)
}

func (input VisionReviewInput) Validate() error {
	if err := input.validateShape(); err != nil {
		return err
	}
	hash, err := visionReviewInputHash(input)
	if err != nil || hash != input.Subject.InputHash {
		return errors.New("Vision Review input hash has drifted")
	}
	return nil
}

func (input VisionReviewInput) validateShape() error {
	if input.ContractID != VisionReviewInputContractID || input.Subject.Validate() != nil || input.BriefCandidate.ValidateFor(input.BriefInput) != nil ||
		input.Admission.ValidateInternalReview() != nil || ValidateVisionReviewAttachments(input.Subject, input.Attachments) != nil {
		return errors.New("invalid Vision Review input")
	}
	brief, subject := input.BriefInput, input.Subject
	if !slices.Contains([]string{"character_identity_anchor", "location_board", "prop_sheet"}, subject.TargetKind) || len(brief.DependencySelections) != 0 {
		return errors.New("Vision Review dependency assets are not ready")
	}
	if subject.WorkspaceID != brief.WorkspaceID || subject.ProjectID != brief.ProjectID || subject.TargetKind != brief.TargetKind || subject.GenerationRound != 1 {
		return errors.New("Vision Review Brief scope has drifted")
	}
	raw, err := json.Marshal(input.BriefCandidate)
	if err != nil {
		return err
	}
	hash, err := canonical.Hash(raw)
	if err != nil || hash != input.BriefContentHash {
		return errors.New("Vision Review Brief content has drifted")
	}
	return input.VisualContext.validateFor(subject.TargetKind)
}

func (visual VisionReviewVisualContext) validateFor(kind string) error {
	g := visual.VisualGrammar
	for _, value := range []string{g.Medium, g.Realism, g.ShapeLanguage, g.Proportion, g.Palette, g.Linework, g.Texture, g.MaterialRendering, g.Lighting, g.Contrast, g.Composition, g.Camera, visual.QCPolicyRef.Owner, visual.QCPolicyRef.Key} {
		if value == "" || strings.TrimSpace(value) != value {
			return errors.New("invalid Vision Review visual context text")
		}
	}
	if !validVisualFoundationMode(visual.ApplicationMode) || !hashPattern.MatchString(visual.ProductionWorldOwnerSetHash) ||
		!hashPattern.MatchString(visual.QCPolicyRef.ContentHash) || visual.PurposeProfile.TargetKind != kind ||
		!reflect.DeepEqual(visual.FidelityInvariants, visualFoundationFidelityInvariants) || validateVisualFoundationPolicy(visual.StylePolicy) != nil ||
		visual.WorldAdaptations == nil || visual.WorldConflicts == nil || visual.CreativeFillProposals == nil ||
		validateWorldAdaptationProposals(visual.WorldAdaptations) != nil || validateVisualWorldConflicts(visual.WorldConflicts) != nil || validateCreativeFillProposals(visual.CreativeFillProposals) != nil ||
		visual.ApplicationMode == "faithful" && len(visual.WorldAdaptations) != 0 {
		return errors.New("invalid Vision Review visual policy")
	}
	for _, values := range [][]string{g.NegativeConstraints, visual.PurposeProfile.DesignFocus, visual.PurposeProfile.ForbiddenChanges} {
		if len(values) == 0 || !sortedUnique(values) {
			return errors.New("Vision Review visual rules must be explicit and ordered")
		}
		for _, value := range values {
			if value == "" || strings.TrimSpace(value) != value {
				return errors.New("invalid Vision Review visual rule")
			}
		}
	}
	return nil
}

func visionReviewInputHash(input VisionReviewInput) (string, error) {
	raw, err := json.Marshal(input)
	if err != nil {
		return "", err
	}
	var material map[string]json.RawMessage
	if err = json.Unmarshal(raw, &material); err != nil {
		return "", err
	}
	var subject map[string]json.RawMessage
	if err = json.Unmarshal(material["subject"], &subject); err != nil {
		return "", err
	}
	delete(subject, "input_hash")
	material["subject"], err = json.Marshal(subject)
	if err != nil {
		return "", err
	}
	raw, err = json.Marshal(material)
	if err != nil {
		return "", err
	}
	return canonical.Hash(raw)
}

func DecodeVisionReviewInput(raw json.RawMessage) (VisionReviewInput, error) {
	var input VisionReviewInput
	if err := canonical.Decode(raw, &input); err != nil {
		return VisionReviewInput{}, err
	}
	if err := input.Validate(); err != nil {
		return VisionReviewInput{}, err
	}
	wire, err := canonical.JSON(raw)
	if err != nil {
		return VisionReviewInput{}, err
	}
	typed, err := json.Marshal(input)
	if err != nil {
		return VisionReviewInput{}, err
	}
	typedWire, err := canonical.JSON(typed)
	if err != nil || !bytes.Equal(wire, typedWire) {
		return VisionReviewInput{}, errors.New("Vision Review input wire shape is incomplete")
	}
	return input, nil
}
